src/goto-instrument/contracts/dynamic-frames/dfcc_utils.cpp

```diff
@@ -0,0 +1,577 @@
+              << messaget::eom;
+  const auto &size =
+    size_of_expr(expr.type(), namespacet(goto_model.symbol_table));
+
+  PRECONDITION_WITH_DIAGNOSTICS(
+    size.has_value(),
+    "dfcc_utilst::make_sizeof_expr: no definite size for lvalue target:\n" +
+      expr.pretty());
+
+  return typecast_exprt::conditional_cast(size.value(), size_type());
+}
+
+exprt dfcc_utilst::make_map_start_address(const exprt &expr)
+{
+  return typecast_exprt::conditional_cast(
+    address_of_exprt(index_exprt(expr, from_integer(0, c_index_type()))),
+    pointer_type(bool_typet()));
+}
+
+void dfcc_utilst::inline_function(const irep_idt &function_id)
+{
+  auto &goto_function = goto_model.goto_functions.function_map.at(function_id);
+
+  PRECONDITION_WITH_DIAGNOSTICS(
+    goto_function.body_available(),
+    "dfcc_utilst::inline_function: '" + id2string(function_id) +
+      "' must have a body");
+
+  inlining_decoratort decorated(log.get_message_handler());
+  namespacet ns{goto_model.symbol_table};
+  goto_function_inline(
+    goto_model.goto_functions, function_id, ns, log.get_message_handler());
+
+  decorated.throw_on_recursive_calls(log, 0);
+  decorated.throw_on_no_body(log, 0);
+  decorated.throw_on_missing_function(log, 0);
+  decorated.throw_on_not_enough_arguments(log, 0);
+
+  goto_model.goto_functions.update();
+}
+
+void dfcc_utilst::inline_function(
+  const irep_idt &function_id,
+  std::set<irep_idt> &no_body,
+  std::set<irep_idt> &recursive_call,
+  std::set<irep_idt> &missing_function,
+  std::set<irep_idt> &not_enough_arguments)
+{
+  auto &goto_function = goto_model.goto_functions.function_map.at(function_id);
+
+  PRECONDITION_WITH_DIAGNOSTICS(
+    goto_function.body_available(),
+    "dfcc_utilst::inline_function: '" + id2string(function_id) +
+      "' must have a body");
+
+  inlining_decoratort decorated(log.get_message_handler());
+  namespacet ns{goto_model.symbol_table};
+  goto_function_inline(
+    goto_model.goto_functions, function_id, ns, log.get_message_handler());
+  no_body.insert(
+    decorated.get_no_body_set().begin(), decorated.get_no_body_set().end());
+  recursive_call.insert(
+    decorated.get_recursive_call_set().begin(),
+    decorated.get_recursive_call_set().end());
+  missing_function.insert(
+    decorated.get_missing_function_set().begin(),
+    decorated.get_missing_function_set().end());
+  not_enough_arguments.insert(
+    decorated.get_not_enough_arguments_set().begin(),
+    decorated.get_not_enough_arguments_set().end());
+  goto_model.goto_functions.update();
+}
+
+void dfcc_utilst::check_loop_freeness(
+  const irep_idt &function_id,
+  std::string msg,
+  int err)
+{
+  auto &goto_function = goto_model.goto_functions.function_map.at(function_id);
+  if(!is_loop_free(goto_function.body, ns, log))
+  {
+    log.error() << "loop-freeness check failed for function '"
+                << id2string(function_id) << "': " << msg << messaget::eom;
+    throw err;
+  }
+}
```

Comment:
I'd suggest to return a `bool`, mark the function `NODISCARD`, and have the caller throw whatever exception is most appropriate.

Reply:
fixed